### PR TITLE
[WIP] Have a same tag-name to define publish artifact configurations.

### PR DESCRIPTION
**Pull request by bdpiparva**
Use same tags to configure build, test or pluggable artifact.

#### Examples
```xml
<artifacts>
  <artifact type="build" src="build/libs" dest="dist"/>
  <artifact type="test" src="test-results" dest="test"/>
  <artifact type="plugin" id="installer" storeId="docker">
    <property>
        <key>BuildFile</key>
        <value>docker-build.json</value>
    </property>
  </artifact>
</artifacts>
```

### CODE REVIEW

**Comment by arvindsv** [comment]
@varshavaradarajan has been chasing me to agree to adding a pluginId next to storeId, to make it easier to use with templates (to be able to parameterize storeId and still be able to show a UI). Might need to consider that too.

**Comment by varshavaradarajan** [comment]
@arvindsv , that's another PR and its pending discussion with @jyotisingh - https://github.com/gocd/gocd/issues/4206. I was just trying to find arguments against the default param thing so that we don't delve into it only to find out later that we can't do it. Apparently you think I have strong opinions. I am just a lowly developer.

**Comment by arvindsv** [comment]
Ha ha! :) Tsk, tsk.

You made good points, though, in arguing for it.

File: config/config-api/src/main/java/com/thoughtworks/go/config/ArtifactConfig.java [hunk 1]
```diff
@@ -17,111 +17,204 @@
...
         return errors().isEmpty();
     }
 
+    @Override
     public void validate(ValidationContext validationContext) {
-        if (!StringUtils.isBlank(destination) && (!(destination.equals(DEFAULT_ROOT.getPath()) || new FilePathTypeValidator().isPathValid(destination)))) {
+        if (artifactType == ArtifactType.plugin) {
+            validatePluggableArtifact(validationContext);
+        } else {
+            validateBuiltInArtifact(validationContext);
```
Comment (varshavaradarajan) [inline config/config-api/src/main/java/com/thoughtworks/go/config/ArtifactConfig.java]: Add more validations - 
1. If id and store id are given, src and dest must not be given
2. If src and destination are provided, pluggable artifact properties, id and store id must not be given.

**Pull request closed without merging**